brainiak/reprsimil/brsa.py

```diff
@@ -867,38 +1324,258 @@ def _fit_RSA_UV(self, X, Y, X0,
             if GP_inten:
                 est_intensity_kernel_r = np.exp(current_GP[1] / 2.0)
                 K_major = np.exp(- (dist2 / est_space_smooth_r**2 +
-                                 inten_diff2 / est_intensity_kernel_r**2)
+                                    inten_diff2 / est_intensity_kernel_r**2)
                                  / 2.0)
-                K = K_major + np.diag(np.ones(n_V) * self.eta)
-                est_std_log_SNR = (np.dot(current_logSNR2, np.dot(
-                    np.linalg.inv(K), current_logSNR2)) / n_V / 4)**0.5
-                # divided by 4 because we used
-                # log(SNR^2) instead of log(SNR)
-                return est_cov_AR1_UV, estU_chlsk_l_AR1_UV, est_SNR_AR1_UV, \
-                    est_beta_AR1_UV, est_beta0_AR1_UV, est_sigma_AR1_UV, \
-                    est_rho1_AR1_UV, est_space_smooth_r, \
-                    est_std_log_SNR, est_intensity_kernel_r
-            # When GP_inten is True, the following lines won't be reached
             else:
+                est_intensity_kernel_r = None
                 K_major = np.exp(- dist2 / est_space_smooth_r**2 / 2.0)
-                K = K_major + np.diag(np.ones(n_V) * self.eta)
-                est_std_log_SNR = (np.dot(current_logSNR2, np.dot(
-                    np.linalg.inv(K), current_logSNR2)) / n_V / 4)**0.5
-                return est_cov_AR1_UV, estU_chlsk_l_AR1_UV, est_SNR_AR1_UV, \
-                    est_beta_AR1_UV, est_beta0_AR1_UV, \
-                    est_sigma_AR1_UV, est_rho1_AR1_UV, est_space_smooth_r, \
-                    est_std_log_SNR
+            K = K_major + np.diag(np.ones(n_V) * self.eta)
+            invK_tilde_log_SNR = np.linalg.solve(K, current_logSNR2) / 2
+            log_SNR_invK_tilde_log_SNR = np.dot(current_logSNR2,
+                                                invK_tilde_log_SNR) / 2
+            tau2, _ = self.tau2_prior(log_SNR_invK_tilde_log_SNR, n_V,
+                                      self.tau_range)
+            est_std_log_SNR = tau2 ** 0.5
+        else:
+            est_space_smooth_r = None
+            est_intensity_kernel_r = None
+            est_std_log_SNR = None
+        return est_cov_AR1_UV, estU_chlsk_l_AR1_UV, est_SNR_AR1_UV, \
+            est_beta_AR1_UV, est_beta0_AR1_UV, est_beta_AR1_UV_latent,\
+            est_sigma_AR1_UV, est_rho1_AR1_UV, est_space_smooth_r, \
+            est_std_log_SNR, est_intensity_kernel_r, X0
+
+    def _transform(self, Y, scan_onsets, beta, beta0,
+                   rho_e, sigma_e, rho_X, sigma2_X, rho_X0, sigma2_X0):
+        """ Given the data Y and the response amplitudes beta and beta0
+            estimated in the fit step, estimate the corresponding X and X0.
+            It is done by a forward-backward algorithm.
+            We assume X and X0 both are vector autoregressive (VAR)
```

Comment:
The reason is practically the intrinsic fluctuation in fMRI and the design matrix are often smooth. For the purpose of decoding, estimating the AR(1) coefficients of X and X0 obtained in training data and use it (together with the variance parameter) as prior can reduce variance in the transforming step. I do not know how to put AR(1) constraint on X0 in fitting step. But they usually come out with autocorrelation automatically if the data has this property. Because X0 is estimated from residuals of so many voxels, I think the data itself provide sufficient constraint without additionally imposing prior during fitting.